sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -119,26 +119,60 @@ class CacheManager extends Logging {
     while (it.hasNext) {
       val cd = it.next()
       if (cd.plan.find(_.sameResult(plan)).isDefined) {
-        cd.cachedRepresentation.cachedColumnBuffers.unpersist(blocking)
+        cd.cachedRepresentation.clearCache(blocking)
         it.remove()
       }
     }
   }
 
+  /**
+   * Materialize the cache that refers to the given physical plan.
```

Comment:
why do we do this?

Reply:
The current approach of this pr is; `cache()` just registers an entry without building a `RDD` in `CacheManager` and then `InMemoryTableScanExec` re-registers an entry to build (materialize) a `RDD` in `CacheManager`. So, I added this function for  `InMemoryTableScanExec`  to re-register these entries in  `CacheManager`. But, I don't think this is the best, so I'd like to have any suggestion.

Reply:
Why do we need to involve `CacheManager`? Shall we just make the creation of RDD lazy in `InMemoryRelation` and trigger the materialization in `InMemoryTableScanExec`?

Reply:
I thought, since `InMemoryRelation` was copied in a tree sometimes, the lazy update of `_cachedColumnBuffers` always didn't lead to the  materialization of the corresponding cache entry in `CacheManager` (maybe...).  If so, following queries might have unnecessary matiralization repeatedly? Therefore, I though we needed to directly update the entry in `CacheManager`.

Reply:
something like
```
class InMemoryRelation(private var _cachedColumnBuffers: RDD[CachedBatch] = null) {
  def cachedColumnBuffers = {
    if (_cachedColumnBuffers == null) {
      synchronized {
        if (_cachedColumnBuffers == null) {
          _cachedColumnBuffers = buildBuffer()
        }
      }
    }
    _cachedColumnBuffers
  }
}

```

Reply:
once it's materialized, it's still materialized after copy

Reply:
I checked the suggested, but some queries didn't work well;
the query in the description was ok, but a query below wrongly cached two different RDDs (I checked the `Strorage` tab in the web UI);
```
scala> sql("SET spark.sql.crossJoin.enabled=true")
scala> val df = spark.range(100000000L).cache()
scala> df.join(df).show
```
This is because Analyzer copies an `InMemoryRelation` (`_cachedColumnBuffers` = null) node via `newInstance` then they build RDDs, respectively. Thought?

Reply:
```
class CachedRDDBuilder(private var _cachedColumnBuffers: RDD[CachedBatch] = null) {
  def cachedColumnBuffers = {
    if (_cachedColumnBuffers == null) {
      synchronized {
        if (_cachedColumnBuffers == null) {
          _cachedColumnBuffers = buildBuffer()
        }
      }
    }
    _cachedColumnBuffers
  } 
}

class InMemoryRelation(cacheBuilder: CachedRDDBuilder = new CachedRDDBuilder()) {
  // newInstance should keep the existing CachedRDDBuilder
  def newInstance()...
}
```

then in the physical plan and cache manager, just call `relation.cacheBuilder.cachedColumnBuffers`

Reply:
aha, I'll fix that way. Thanks!